### ISSUE TITLE
Time-varying and uncertain generation times

### ISSUE THREAD

**Issue by seabbs** [epinowcast/epinowcast#178]
In https://github.com/epinowcast/epinowcast/pull/152 a deterministic renewal process was added as an optional generative process. As it currently stands this is the most basic possible implementation with the reporting distribution being assumed to be static and being passed in as a pre-calculated PMF.

To match other packages (i.e EpiNow2) we likely want to support uncertainty and time-varying/group-varying distributions. As @adrian-lison notes we may also want to consider the various forms of stochastic implementaton in order to pick one that is best performing (see https://github.com/epinowcast/epinowcast/pull/152#issuecomment-1333714695)

In addition to passing in more complex distributions, we may wish to extend the complexity of the underlying model to fit changes in the generation time. Identification of these changes can be difficult without additional sources of data so this may need some thought but in theory, the discrete-time hazard machinery we use for the observed reporting distributions should be easily repurposed for this task. We may need to consider this carefully as a forward generative process (i.e using the case reproduction number) may be easier to link to changes in NPIs. 

Its important to note that in order to maintain usability it is key the user interface for this functionality is as simple as possible and that highly generic but complex specification is not the first thing users come in to contact with when exploring this functionality.

Something else to keep in mind is how much fitting of this distribution we would like to do in model vs allowign users to pass in complex posteriors as priors obtained using other data sources. 

Note any development work on this should target develop. See inst/examples/germany_latent_renewal.R for an example of the current functionality.

Flagging @adrian-lison and @parksw3 who are likely interested in thinking about this. We may also want to move this discussion to community.epinowcast.org for now in order to maximise visibility

**Comment by parksw3** [comment]
> In addition to passing in more complex distributions, we may wish to extend the complexity of the underlying model to fit changes in the generation time. Identification of these changes can be difficult without additional sources of data

I think this is impossible in theory (i.e., the incidence curve doesn't provide any information about the underlying GI or how it changes). In practice, it can be possible with strong priors (e.g., given constraints on R0 and a range of possible changes in Reff over a time step, you could attribute the remaining changes to the GI). But this would be largely prior driven rather than data.

> We may need to consider this carefully as a forward generative process (i.e using the case reproduction number) may be easier to link to changes in NPIs.

Using the case reproduction number is probably more difficult to link to changes in NPIs because you have to know what's happening after time t. And you need to link that with the forward generation interval, which is constantly changing as long as conditions are changing. I would think that the instantaneous formulation (the standard formulation) would be easier given that the instantaneous GI is insensitive to strength-like changes (e.g., reduction in contact rates, vaccination, etc). The hazard-based formulation also fits nicely into the instantaneous formulation.

Sorry, I said I'll write about this in community.epinowcast.org a long time ago... Meanwhile, different renewal equation formulations (case vs instantaneous), hazard, etc are described in https://github.com/parksw3/individual_intervention...

**Comment by adrian-lison** [comment]
I guess we should only start working on this after we have the option to (partially) decouple the nowcasting and the renewal component, since we will likely need a lot of data to have any hope of recovering something - in particular **fully reported** data.

What I like is your idea @seabbs to extend the capabilities for specifying priors on the generation interval etc. A first valuable extension would already be to allow the user to provide fixed distributions that vary over time, so that the model can be fit on time series with changing generation times (NPIs, new strains etc) if there is external knowledge on these changes.

**Comment by adrian-lison** [comment]
Okay, I now see that Sam is one step ahead again and implemenred the latter already in #183 :D

**Comment by seabbs** [comment]
Yes agree @parksw3 it is impossible in theory but from a workbench mentality we can provide a general model where set specifications are identifiable and some are not (we can then work on that going forward with novel data etc.). 

> Using the case reproduction number is probably more difficult to link to changes

Yes agree. This was a moment of madness. 

>  I guess we should only start working on this after we have the option to (partially) decouple the nowcasting and the renewal component, since we will likely need a lot of data to have any hope of recovering something - in particular fully reported data.

Agree the complex case is not a priority right now. 

> Okay, I now see that Sam is one step ahead again and implemenred the latter already in https://github.com/epinowcast/epinowcast/pull/183 :D

So I have only currently done this for the latent reporting distribution but I was planning on adding this in for the generation time as well (or anyone else who would like to take a crack can!) before we release `0.2.0` because as you say it is a nice bit of functionality to have and doesn't really cost us that much.
